rapid_pro_tools/rapid_pro_client.py

```diff
@@ -26,6 +29,68 @@ def __init__(self, server, token):
         """
         self.rapid_pro = TembaClient(server, token)
         
+    def list_archives(self, archive_type=None):
+        """
+        Lists all of the available archives on this instance.
+
+        Returns a list of objects with archive metadata. Pass one of these metadata objects into
+        `RapidProClient.get_archive` to download the archive itself. Note that the download links in the returned
+        metadata are only valid for a short period of time. From experimentation as at January 2020,
+        links are valid for 1 day after this call is made.
+        
+        :param archive_type: The type of archives to list (either 'message' or 'run') or None.
+                             If None, lists both types of archive.
+        :type archive_type: str | None
+        :return: List of available archives on this instance.
+        :rtype: list of temba_client.v2.types.Archive
+        """
+        assert archive_type in {"message", "run"}
+
+        return self.rapid_pro.get_archives(archive_type=archive_type).all(retry_on_rate_exceed=True)
+
+    def get_archive(self, archive_metadata):
+        """
+        Downloads the archive specified by an archive metadata object, and converts it into a valid list of Message
+        or Run objects.
+        
+        Note: Deserializing to message objects is not yet implemented, so requests for message archives will fail
+        with an assertion error.
+        TODO: Support deserializing messages
+
+        :param archive_metadata: Metadata for the archive. To obtain these, see `RapidProClient.list_archives`.
+        :type archive_metadata: temba_client.v2.types.Archive
+        :return: Data downloaded from the archive.
+        :rtype: list of temba_client.v2.Message | list of temba_client.v2.Run
+        """
+        # Download the archive, which is in a gzipped JSONL format, and decompress.
+        log.info(f"Downloading archive from {archive_metadata.download_url}")
+        archive_response = urllib.request.urlopen(archive_metadata.download_url)
+        raw_file = BytesIO(archive_response.read())
+        decompressed_file = gzip.GzipFile(fileobj=raw_file)
```

Comment:
Doing this in memory could use a lot of ram - presumably you've tested this inside a reasonably configured Docker container?

Reply:
While the memory usage is likely to be high by normal standards, it should be pretty low compared to what the rest of the pipelines need. The largest whole-month archive for IMAQAL is 60MB compressed, so at most a few GB decompressed. I'm not having any problems so far, but will keep this in mind and come back to this if we need to change.

Reply:
Peak memory use approx. 1GB when loading all the records for CSAP